metabase/metabase
# Fix unable to start on windows with h2

We have a special treatment for h2 connections in that every time we try to get a connection from data-source, we check whether or not we should update h2 version, this is a trick we did in https://github.com/metabase/metabase/pull/26576 during h2 migration from v1 to v2.

The function that is used to check whether or not we need to update h2 is [metabase.db.update-h2/update-if-needed!](https://github.com/metabase/metabase/blob/9413706bd2778dedf575eb38ce9a7de06253c96f/src/metabase/db/update_h2.clj#L90), the basic idea of this function is that: if the current connection is an h2 database, open the file and read the header to see if the database is v1, if it is, then try to migrate the database to version 2.

This causes a deadlock while trying to run the `migrations/001_update_migrations.yaml::v46.00-086` migration on Windows with this error message
> Caused by: java.io.IOException: The process cannot access the file because another process has locked a portion of the file

Why does this happen on Windows but not on our CI or Mac? I don't know for sure... it might be that liquibase locked the h2 file but did not release the lock properly.

One thing I know for sure is that we _shouldn't_ do the upgrade check by reading h2 file every time a connection is fetched, this is super inefficient. And also I do not think that we need to do this in `.getConnection` we can just do that when we create a data source. I think the original idea of why we do that during `getConnection` is that we do it "on demand", but that's unnecessary IMO.

How to test
1. If you're on windows: download the jar and try to run it
2. if you're Mac, follow this [tutorial](https://mac.getutm.app/gallery/windows-11-arm) using UTM to start a Windows VM, then download a jar from this branch and start it. You shouldn't see any errors.
3. If you're on ubuntu: you'll need to find some VM yourself bc I really don't know... or maybe get a mac? :p 

Fixes https://github.com/metabase/metabase/issues/39040
